File: claude-code-os-implementation/03-ai-growth-engine/organic-marketing-package/content-agents/api/routes/aeo.py
"""
AEO (Answer Engine Optimization) content generation routes.

This module defines API endpoints for AEO content generation using the AEOAgent.
"""

from fastapi import APIRouter, Depends, HTTPException
from typing import Dict, Any, List, Optional
from datetime import datetime
from pathlib import Path
import logging
import time

from api.dependencies import get_request_id
from api.models import (
    ContentResponse,
    ContentMetadata,
    ErrorResponse,
    FAQGenerationRequest,
    FAQSchemaRequest,
    ProductSchemaRequest,
    AIOptimizedContentRequest,
    ComparisonContentRequest,
    FAQContentResponse,
    SchemaResponse
)
from agents.aeo_agent import AEOAgent

# Configure logging
logger = logging.getLogger(__name__)

# Create router
router = APIRouter(
    prefix="/aeo",
    tags=["aeo"],
    responses={
        500: {"model": ErrorResponse, "description": "Internal server error"}
    }
)


# ============================================================================
# Endpoints
# ============================================================================

@router.post(
    "/generate-faq",
    response_model=FAQContentResponse,
    summary="Generate FAQ content",
    description="Generate FAQ content optimized for AI assistant citation"
)
async def generate_faq_content(
    request: FAQGenerationRequest,
    request_id: str = Depends(get_request_id)
) -> Dict[str, Any]:
    """
    Generate FAQ content.

    Args:
        request: FAQ generation request
        request_id: Unique request identifier

    Returns:
        FAQ content response with generated content and metadata

    Raises:
        HTTPException: If content generation fails
    """
    logger.info(f"[{request_id}] Generating FAQ content: topic='{request.topic}', num_questions={request.num_questions}")
    start_time = time.time()

    try:
        # Initialize agent
        agent = AEOAgent()

        # Generate FAQ content
        content, file_path = agent.generate_faq_content(
            topic=request.topic,
            num_questions=request.num_questions,
            target_audience=request.target_audience,
            include_product_mentions=request.include_product_mentions
        )

        # Calculate generation time
        generation_time_ms = int((time.time() - start_time) * 1000)

        # Build response
        response = {
            "request_id": request_id,
            "content": content,
            "file_path": str(file_path),
            "metadata": {
                "agent": "aeo_agent",
                "model": "claude-sonnet-4-5-20250929",
                "tokens_used": 0,  # TODO: Implement token tracking
                "generation_time_ms": generation_time_ms,
                "timestamp": datetime.utcnow()
            },
            "status": "success"
        }

        logger.info(f"[{request_id}] Successfully generated FAQ content in {generation_time_ms}ms")
        return response

    except Exception as e:
        logger.error(f"[{request_id}] Error generating FAQ content: {e}", exc_info=True)
        raise HTTPException(
            status_code=500,
            detail={
                "error": "FAQGenerationError",
                "message": f"Failed to generate FAQ content: {str(e)}",
                "request_id": request_id,
                "timestamp": datetime.utcnow().isoformat() + "Z"
            }
        )


@router.post(
    "/generate-faq-schema",
    response_model=SchemaResponse,
    summary="Generate FAQ schema",
    description="Generate JSON-LD FAQ schema markup for AI parsing"
)
async def generate_faq_schema(
    request: FAQSchemaRequest,
    request_id: str = Depends(get_request_id)
) -> Dict[str, Any]:
    """
    Generate FAQ schema.

    Args:
        request: FAQ schema generation request
        request_id: Unique request identifier

    Returns:
        Schema response with generated JSON-LD and metadata

    Raises:
        HTTPException: If schema generation fails
    """
    logger.info(f"[{request_id}] Generating FAQ schema: num_items={len(request.faq_items)}")
    start_time = time.time()

    try:
        # Initialize agent
        agent = AEOAgent()

        # Generate FAQ schema
        schema = agent.generate_faq_schema(faq_items=request.faq_items)

        # Calculate generation time
        generation_time_ms = int((time.time() - start_time) * 1000)

        # Build response
        response = {
            "request_id": request_id,
            "schema": schema,
            "schema_type": "FAQPage",
            "metadata": {
                "agent": "aeo_agent",
                "model": "claude-sonnet-4-5-20250929",
                "tokens_used": 0,  # No tokens used for schema generation
                "generation_time_ms": generation_time_ms,
                "timestamp": datetime.utcnow()
            },
            "status": "success"
        }

        logger.info(f"[{request_id}] Successfully generated FAQ schema in {generation_time_ms}ms")
        return response

    except Exception as e:
        logger.error(f"[{request_id}] Error generating FAQ schema: {e}", exc_info=True)
        raise HTTPException(
            status_code=500,
            detail={
                "error": "FAQSchemaGenerationError",
                "message": f"Failed to generate FAQ schema: {str(e)}",
                "request_id": request_id,
                "timestamp": datetime.utcnow().isoformat() + "Z"
            }
        )


@router.post(
    "/generate-product-schema",
    response_model=SchemaResponse,
    summary="Generate Product schema",
    description="Generate JSON-LD Product schema markup for AI parsing"
)
async def generate_product_schema(
    request: ProductSchemaRequest,
    request_id: str = Depends(get_request_id)
) -> Dict[str, Any]:
    """
    Generate Product schema.

    Args:
        request: Product schema generation request
        request_id: Unique request identifier

    Returns:
        Schema response with generated JSON-LD and metadata

    Raises:
        HTTPException: If schema generation fails
    """
    product_name = request.product_data.get('name', 'Unknown Product')
    logger.info(f"[{request_id}] Generating Product schema: product='{product_name}'")
    start_time = time.time()

    try:
        # Initialize agent
        agent = AEOAgent()

        # Generate Product schema
        schema = agent.generate_product_schema(product_data=request.product_data)

        # Calculate generation time
        generation_time_ms = int((time.time() - start_time) * 1000)

        # Build response
        response = {
            "request_id": request_id,
            "schema": schema,
            "schema_type": "Product",
            "metadata": {
                "agent": "aeo_agent",
                "model": "claude-sonnet-4-5-20250929",
                "tokens_used": 0,  # No tokens used for schema generation
                "generation_time_ms": generation_time_ms,
                "timestamp": datetime.utcnow()
            },
            "status": "success"
        }

        logger.info(f"[{request_id}] Successfully generated Product schema in {generation_time_ms}ms")
        return response

    except Exception as e:
        logger.error(f"[{request_id}] Error generating Product schema: {e}", exc_info=True)
        raise HTTPException(
            status_code=500,
            detail={
                "error": "ProductSchemaGenerationError",
                "message": f"Failed to generate Product schema: {str(e)}",
                "request_id": request_id,
                "timestamp": datetime.utcnow().isoformat() + "Z"
            }
        )


@router.post(
    "/generate-ai-content",
    response_model=FAQContentResponse,
    summary="Generate AI-optimized content",
    description="Generate content specifically optimized for AI assistant citation"
)
async def generate_ai_optimized_content(
    request: AIOptimizedContentRequest,
    request_id: str = Depends(get_request_id)
) -> Dict[str, Any]:
    """
    Generate AI-optimized content.

    Args:
        request: AI-optimized content generation request
        request_id: Unique request identifier

    Returns:
        Content response with generated content and metadata

    Raises:
        HTTPException: If content generation fails
    """
    logger.info(f"[{request_id}] Generating AI-optimized content: question='{request.question}', type={request.content_type}")
    start_time = time.time()

    try:
        # Initialize agent
        agent = AEOAgent()

        # Generate AI-optimized content
        content, file_path = agent.generate_ai_optimized_content(
            question=request.question,
            content_type=request.content_type,
            include_sources=request.include_sources
        )

        # Calculate generation time
        generation_time_ms = int((time.time() - start_time) * 1000)

        # Build response
        response = {
            "request_id": request_id,
            "content": content,
            "file_path": str(file_path),
            "metadata": {
                "agent": "aeo_agent",
                "model": "claude-sonnet-4-5-20250929",
                "tokens_used": 0,  # TODO: Implement token tracking
                "generation_time_ms": generation_time_ms,
                "timestamp": datetime.utcnow()
            },
            "status": "success"
        }

        logger.info(f"[{request_id}] Successfully generated AI-optimized content in {generation_time_ms}ms")
        return response

    except Exception as e:
        logger.error(f"[{request_id}] Error generating AI-optimized content: {e}", exc_info=True)
        raise HTTPException(
            status_code=500,
            detail={
                "error": "AIContentGenerationError",
                "message": f"Failed to generate AI-optimized content: {str(e)}",
                "request_id": request_id,
                "timestamp": datetime.utcnow().isoformat() + "Z"
            }
        )


@router.post(
    "/generate-comparison",
    response_model=FAQContentResponse,
    summary="Generate comparison content",
    description="Generate comparison content optimized for 'best' and 'vs' queries"
)
async def generate_comparison_content(
    request: ComparisonContentRequest,
    request_id: str = Depends(get_request_id)
) -> Dict[str, Any]:
    """
    Generate comparison content.

    Args:
        request: Comparison content generation request
        request_id: Unique request identifier

    Returns:
        Content response with generated comparison and metadata

    Raises:
        HTTPException: If content generation fails
    """
    logger.info(f"[{request_id}] Generating comparison content: topic='{request.comparison_topic}', items={len(request.items_to_compare)}")
    start_time = time.time()

    try:
        # Initialize agent
        agent = AEOAgent()

        # Generate comparison content
        content, file_path = agent.generate_comparison_content(
            comparison_topic=request.comparison_topic,
            items_to_compare=request.items_to_compare,
            include_recommendation=request.include_recommendation
        )

        # Calculate generation time
        generation_time_ms = int((time.time() - start_time) * 1000)

        # Build response
        response = {
            "request_id": request_id,
            "content": content,
            "file_path": str(file_path),
            "metadata": {
                "agent": "aeo_agent",
                "model": "claude-sonnet-4-5-20250929",
                "tokens_used": 0,  # TODO: Implement token tracking
                "generation_time_ms": generation_time_ms,
                "timestamp": datetime.utcnow()
            },
            "status": "success"
        }

        logger.info(f"[{request_id}] Successfully generated comparison content in {generation_time_ms}ms")
        return response

    except Exception as e:
        logger.error(f"[{request_id}] Error generating comparison content: {e}", exc_info=True)
        raise HTTPException(
            status_code=500,
            detail={
                "error": "ComparisonGenerationError",
                "message": f"Failed to generate comparison content: {str(e)}",
                "request_id": request_id,
                "timestamp": datetime.utcnow().isoformat() + "Z"
            }
        )


@router.get(
    "/health",
    summary="Health check",
    description="Check if the AEO router is operational"
)
async def health_check() -> Dict[str, str]:
    """
    Health check endpoint.

    Returns:
        Health status
    """
    return {
        "status": "healthy",
        "service": "aeo",
        "timestamp": datetime.utcnow().isoformat() + "Z"
    }

File: claude-code-os-implementation/03-ai-growth-engine/organic-marketing-package/packages/dashboard/src/lib/api/aeo.ts
/**
 * AEO (Answer Engine Optimization) API Client
 *
 * Provides typed functions for interacting with the AEO backend endpoints
 * including citation tracking, content optimization, and metrics.
 */

import { apiClient } from '@/lib/api-client'
import type { ApiClientOptions } from '@/lib/api-client'

// ============================================================================
// Types
// ============================================================================

export interface ContentMetadata {
  agent: string
  model: string
  tokens_used: number
  generation_time_ms: number
  timestamp: string
}

export interface CitationRecord {
  tracking_id: string
  platform: string
  query: string
  brand_mentioned: boolean
  citation_url: string | null
  citation_context: string | null
  opportunity_score: number
  query_category: string | null
  created_at: string | null
}

export interface PlatformStats {
  total_queries: number
  brand_mentioned: number
  citation_rate: number
  avg_opportunity_score?: number
  top_missed_queries?: string[]
}

export interface CitationStatusResponse {
  request_id: string
  summary: {
    total_tracked: number
    brand_mentioned: number
    citation_rate_percent: number
    avg_opportunity_score: number
    period_days: number
  }
  platforms: Record<string, PlatformStats>
  records: CitationRecord[]
  metadata: {
    generation_time_ms: number
    timestamp: string
  }
  status: string
}

export interface MissedOpportunity {
  query: string
  platform: string
  opportunity_score: number
  competitors_cited: Record<string, unknown> | null
  query_category: string | null
  date: string | null
}

export interface Recommendation {
  id: number
  type: string
  title: string
  description: string
  priority: string
  expected_impact: number | null
  effort: string | null
  ai_platform: string | null
}

export interface CitationAnalysisResponse {
  request_id: string
  brand: string
  analysis_period_days: number
  missed_opportunities: MissedOpportunity[]
  total_missed_opportunities: number
  recommendations: Recommendation[]
  platform_analysis: Record<string, PlatformStats>
  metadata: {
    generation_time_ms: number
    timestamp: string
  }
  status: string
}

export interface CompetitorStats {
  competitor: string
  total_queries: number
  times_mentioned: number
  citation_rate_percent: number
  platform_rates: Record<string, number>
}

export interface CompetitorCitationsResponse {
  request_id: string
  period_days: number
  brand_stats: {
    brand: string
    total_queries: number
    times_mentioned: number
    citation_rate_percent: number
  }
  competitors: CompetitorStats[]
  monitored_competitors: string[]
  metadata: {
    generation_time_ms: number
    timestamp: string
  }
  status: string
}

export interface ContentOptimizeResponse {
  request_id: string
  optimization_analysis: string
  optimization_level: string
  target_queries: string[]
  file_path: string
  metadata: ContentMetadata & { timestamp: string }
  status: string
}

export interface BlogGenerateResponse {
  request_id: string
  content: string
  file_path: string
  topic: string
  schemas: {
    faq?: Record<string, unknown>
    faq_error?: string
  }
  metadata: ContentMetadata & { timestamp: string }
  status: string
}

export interface DashboardOverview {
  citation_rate_percent: number
  citation_rate_change: number
  total_queries_tracked: number
  brand_mentions: number
  avg_opportunity_score: number
  pending_recommendations: number
  period_days: number
}

export interface OpportunityItem {
  query: string
  platform: string
  opportunity_score: number
  query_category: string | null
}

export interface CompetitorOverview {
  name: string
  citation_rate: number
  total_queries: number
}

export interface AEODashboardResponse {
  request_id: string
  overview: DashboardOverview
  platforms: Record<string, PlatformStats>
  categories: Record<string, {
    total: number
    mentioned: number
    citation_rate: number
  }>
  top_opportunities: OpportunityItem[]
  top_competitors: CompetitorOverview[]
  thresholds: {
    high_opportunity: number
    medium_opportunity: number
    alert_citation_drop: number
  }
  metadata: {
    generation_time_ms: number
    timestamp: string
  }
  status: string
}

export interface FAQContentResponse {
  request_id: string
  content: string
  file_path: string
  metadata: ContentMetadata
  status: string
}

export interface SchemaResponse {
  request_id: string
  schema: string
  schema_type: string
  metadata: ContentMetadata
  status: string
}

// ============================================================================
// Request Types
// ============================================================================

export interface CitationAnalyzeRequest {
  queries?: string[]
  platforms?: string[]
  brand_name?: string
}

export interface ContentOptimizeRequest {
  content: string
  target_queries?: string[]
  optimization_level?: 'light' | 'standard' | 'aggressive'
}

export interface BlogGenerateRequest {
  topic: string
  target_queries?: string[]
  word_count?: number
  include_faq_schema?: boolean
  include_product_schema?: boolean
}

export interface FAQGenerationRequest {
  topic: string
  num_questions?: number
  target_audience?: string
  include_product_mentions?: boolean
}

export interface FAQSchemaRequest {
  faq_items: Array<{ question: string; answer: string }>
}

export interface AIOptimizedContentRequest {
  question: string
  content_type?: 'guide' | 'article' | 'comparison' | 'tutorial'
  include_sources?: boolean
}

export interface ComparisonContentRequest {
  comparison_topic: string
  items_to_compare: string[]
  include_recommendation?: boolean
}

// ============================================================================
// API Base URL
// ============================================================================

const API_BASE = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'
const AEO_BASE = `${API_BASE}/api/aeo`

// ============================================================================
// Citation Tracking Functions
// ============================================================================

/**
 * Get current citation tracking status across AI platforms.
 */
export async function getCitationStatus(params?: {
  platform?: string
  days?: number
  limit?: number
}, options?: ApiClientOptions): Promise<CitationStatusResponse> {
  const searchParams = new URLSearchParams()
  if (params?.platform) searchParams.set('platform', params.platform)
  if (params?.days) searchParams.set('days', params.days.toString())
  if (params?.limit) searchParams.set('limit', params.limit.toString())

  const query = searchParams.toString()
  const url = `${AEO_BASE}/citations/track${query ? `?${query}` : ''}`

  return apiClient.get<CitationStatusResponse>(url, {
    timeout: 15000,
    ...options,
  })
}

/**
 * Analyze citation opportunities across AI platforms.
 */
export async function analyzeCitations(
  request: CitationAnalyzeRequest,
  options?: ApiClientOptions
): Promise<CitationAnalysisResponse> {
  return apiClient.post<CitationAnalysisResponse>(
    `${AEO_BASE}/citations/analyze`,
    request,
    { timeout: 30000, ...options }
  )
}

/**
 * Get competitor citation comparison data.
 */
export async function getCompetitorCitations(params?: {
  days?: number
  competitor?: string
  platform?: string
}, options?: ApiClientOptions): Promise<CompetitorCitationsResponse> {
  const searchParams = new URLSearchParams()
  if (params?.days) searchParams.set('days', params.days.toString())
  if (params?.competitor) searchParams.set('competitor', params.competitor)
  if (params?.platform) searchParams.set('platform', params.platform)

  const query = searchParams.toString()
  const url = `${AEO_BASE}/citations/competitors${query ? `?${query}` : ''}`

  return apiClient.get<CompetitorCitationsResponse>(url, {
    timeout: 15000,
    ...options,
  })
}

// ============================================================================
// Content Optimization Functions
// ============================================================================

/**
 * Optimize content for better AI citation rates.
 */
export async function optimizeContent(
  request: ContentOptimizeRequest,
  options?: ApiClientOptions
): Promise<ContentOptimizeResponse> {
  return apiClient.post<ContentOptimizeResponse>(
    `${AEO_BASE}/content/optimize`,
    request,
    { timeout: 60000, ...options }
  )
}

/**
 * Generate a citation-optimized blog post.
 */
export async function generateCitationBlog(
  request: BlogGenerateRequest,
  options?: ApiClientOptions
): Promise<BlogGenerateResponse> {
  return apiClient.post<BlogGenerateResponse>(
    `${AEO_BASE}/blog/generate`,
    request,
    { timeout: 120000, ...options }
  )
}

// ============================================================================
// Content Generation Functions (existing AEO endpoints)
// ============================================================================

/**
 * Generate FAQ content optimized for AI citation.
 */
export async function generateFAQContent(
  request: FAQGenerationRequest,
  options?: ApiClientOptions
): Promise<FAQContentResponse> {
  return apiClient.post<FAQContentResponse>(
    `${AEO_BASE}/generate-faq`,
    request,
    { timeout: 60000, ...options }
  )
}

/**
 * Generate JSON-LD FAQ schema markup.
 */
export async function generateFAQSchema(
  request: FAQSchemaRequest,
  options?: ApiClientOptions
): Promise<SchemaResponse> {
  return apiClient.post<SchemaResponse>(
    `${AEO_BASE}/generate-faq-schema`,
    request,
    { timeout: 15000, ...options }
  )
}

/**
 * Generate JSON-LD Product schema markup.
 */
export async function generateProductSchema(
  productData: Record<string, unknown>,
  options?: ApiClientOptions
): Promise<SchemaResponse> {
  return apiClient.post<SchemaResponse>(
    `${AEO_BASE}/generate-product-schema`,
    { product_data: productData },
    { timeout: 15000, ...options }
  )
}

/**
 * Generate AI-optimized content for a specific question.
 */
export async function generateAIContent(
  request: AIOptimizedContentRequest,
  options?: ApiClientOptions
): Promise<FAQContentResponse> {
  return apiClient.post<FAQContentResponse>(
    `${AEO_BASE}/generate-ai-content`,
    request,
    { timeout: 60000, ...options }
  )
}

/**
 * Generate comparison content optimized for "best" and "vs" queries.
 */
export async function generateComparisonContent(
  request: ComparisonContentRequest,
  options?: ApiClientOptions
): Promise<FAQContentResponse> {
  return apiClient.post<FAQContentResponse>(
    `${AEO_BASE}/generate-comparison`,
    request,
    { timeout: 60000, ...options }
  )
}

// ============================================================================
// Metrics & Dashboard Functions
// ============================================================================

/**
 * Get comprehensive AEO dashboard metrics.
 */
export async function getAEODashboardMetrics(params?: {
  days?: number
}, options?: ApiClientOptions): Promise<AEODashboardResponse> {
  const searchParams = new URLSearchParams()
  if (params?.days) searchParams.set('days', params.days.toString())

  const query = searchParams.toString()
  const url = `${AEO_BASE}/metrics/dashboard${query ? `?${query}` : ''}`

  return apiClient.get<AEODashboardResponse>(url, {
    timeout: 15000,
    ...options,
  })
}

/**
 * Check AEO service health.
 */
export async function checkAEOHealth(
  options?: ApiClientOptions
): Promise<{ status: string; service: string; timestamp: string }> {
  return apiClient.get(`${AEO_BASE}/health`, {
    timeout: 5000,
    ...options,
  })
}

// ============================================================================
// Convenience exports
// ============================================================================

export const aeoApi = {
  // Citation tracking
  getCitationStatus,
  analyzeCitations,
  getCompetitorCitations,

  // Content optimization
  optimizeContent,
  generateCitationBlog,

  // Content generation
  generateFAQContent,
  generateFAQSchema,
  generateProductSchema,
  generateAIContent,
  generateComparisonContent,

  // Metrics
  getAEODashboardMetrics,
  checkAEOHealth,
}
